### PR TITLE
RFC: Local Cache

### PR DESCRIPTION
**Pull request by jdwyah**
if use_local_cache option is true
cache json of API payloads to local file.
load from cache as a tertiary source.

### CODE REVIEW

File: lib/prefab/config_client.rb
Comment (semanticart) [inline lib/prefab/config_client.rb:201]: ✂️

**Review by semanticart — APPROVED** [review]
LGTM once you remove the `puts`

File: lib/prefab/config_client.rb
Comment (jkebinger) [inline lib/prefab/config_client.rb:182]: i've read this a couple times and i'm still not sure what the path here is when there is or is not an XDG_CACHE_HOME - can this be
1) less terse
2) log the cache location the first time its generated

Comment (jdwyah) [reply lib/prefab/config_client.rb]: mm. Test is more readable https://github.com/prefab-cloud/prefab-cloud-ruby/pull/148/files#diff-0316acc4f7bad770d459a6870363ba6411b13e6ad04a2beef589b767894d0af0R101 but yeah, a bit of a long line. I can tweak it.

File: lib/prefab/config_client.rb
Comment (jkebinger) [inline lib/prefab/config_client.rb:190]: This prevents simultaneous writes but doesn't fix the partial read problem
do at least one of
1) f.flock(File:LOCK_SH) on read so the reads are blocked on write completing https://apidock.com/ruby/v2_5_5/File/flock
2) the atomic write (write and mv) (this will still need the write lock)

File: lib/prefab/config_client.rb
Comment (jkebinger) [inline lib/prefab/config_client.rb:196]: The lock prevents multiple writes, but this still needs the atomic-write pattern so non one reads a half written file

File: lib/prefab/config_client.rb
Comment (jdwyah) [reply lib/prefab/config_client.rb:196]: 👍 addressed in loader with LOCK_SH

**Pull request merged**
